src/plugins/scopes/Scope.js

```diff
@@ -0,0 +1,204 @@
+/* @flow */
+
+import type Node from '../../elements/Node';
+import Reference from './Reference';
+import Variable from './Variable';
+import {default as Definition, types, typeOrder} from './Definition';
+
+export default class Scope {
+    constructor({node, parentScope, isProgramScope, isFunctionScope, isClassScope, isArrowFunctionScope}: {
+        node: Node,
+        parentScope: ?Scope,
+        isProgramScope?: boolean,
+        isFunctionScope?: boolean,
+        isClassScope?: boolean,
+        isArrowFunctionScope?: boolean
+    }) {
+        this._node = node;
+        this._parentScope = parentScope;
+        if (parentScope) {
+            parentScope._childScopes.push(this);
+            this._depth = parentScope._depth + 1;
+        } else {
+            this._depth = 0;
+        }
+        this._childScopes = [];
+        this._variables = new Map();
+        this._references = new Map();
+        this._isProgramScope = Boolean(isProgramScope);
+        this._isFunctionScope = Boolean(isFunctionScope);
+        this._isClassScope = Boolean(isClassScope);
+        this._isArrowFunctionScope = Boolean(isArrowFunctionScope);
+    }
+
+    _isProgramScope: boolean;
+    _isFunctionScope: boolean;
+    _isClassScope: boolean;
+    _isArrowFunctionScope: boolean;
+    _node: Node;
+    _depth: number;
+    _parentScope: ?Scope;
+    _childScopes: Scope[];
+    _variables: Map<string, Variable[]>;
+    _references: Map<string, Reference[]>;
+
+    _addVariable(variable: Variable) {
+        let variables = this._variables.get(variable.name);
+        if (variables) {
+            variables.push(variable);
+            variables.sort((variable1: Variable, variable2: Variable) => {
+                let typeOrder1 = typeOrder[variable1._type];
+                let typeOrder2 = typeOrder[variable2._type];
+                return typeOrder1 > typeOrder2 ? 1 : (typeOrder1 < typeOrder2 ? -1 : 0);
+            });
+        } else {
+            this._variables.set(variable.name, [variable]);
```

Comment:
```js
Object.keys(this._variables).map(key => this._variables[key])
```
Would also work, `Map` supposed to be slower, besides `values` are cool, but `Map` have, as you probably know, a reach api, to use it only because of the `values` method